src/oemof/solph/components/_generic_storage.py

```diff
@@ -1229,27 +1241,76 @@ def _old_storage_capacity_rule_end(block):
                             "for a Flow going into or out of "
                             "a GenericStorage unit "
                             "in a multi-period model!"
-                            " Value for {} is missing.".format(n)
+                            f" Value for {n} is missing."
                         )
                         raise ValueError(msg)
+                    # get the period matrix describing the temporal distance
+                    # between all period combinations.
+                    periods_matrix = m.es.periods_matrix
+
+                    # get the index of the minimum value in each row greater
+                    # equal than the lifetime. This value equals the
+                    # decommissioning period if not zero. The index of this
+                    # value represents the investment period. If np.where
+                    # condition is not met in any row, min value will be zero
+                    decomm_periods = np.argmin(
+                        np.where(
+                            (periods_matrix >= lifetime),
+                            periods_matrix,
+                            np.inf,
+                        ),
+                        axis=1,
+                    )
+
+                    # no decommissioning in first period
+                    expr = self.old_end[n, 0] == 0
+                    self.old_rule_end.add((n, 0), expr)
+
+                    # all periods not in decomm_periods have no decommissioning
+                    # zero is excluded
                     for p in m.PERIODS:
-                        # No shutdown in first period
-                        if p == 0:
+                        if p not in decomm_periods and p != 0:
                             expr = self.old_end[n, p] == 0
                             self.old_rule_end.add((n, p), expr)
-                        elif lifetime <= m.es.periods_years[p]:
-                            # Obtain commissioning period
-                            comm_p = 0
-                            for k, v in enumerate(m.es.periods_years):
-                                if m.es.periods_years[p] - lifetime - v < 0:
-                                    # change of sign is detected
-                                    comm_p = k - 1
-                                    break
-                            expr = self.old_end[n, p] == self.invest[n, comm_p]
-                            self.old_rule_end.add((n, p), expr)
+
+                    # multiple invests can decommission in the same period
```

Comment:
Another minor wording remark: I think, it should be "can be decommissioned" not "can decommision".